File: python/desireeia/autoconfig.py
"""Automatic configuration of a model for the current machine.

Mirror of the C# ``AutoConfigurator`` (src/DesireeIA/AutoConfig.cs): same
defaults, same rules, so the NuGet and PyPI packages behave identically.

- backend, threads and RAM budget come from the native planner
  (:func:`build_plan`), which probes CPU features, CUDA/Intel/Metal devices
  and free memory;
- sampling is never greedy: greedy decoding falls into repetitive loops on
  long generations;
- the context is as large as the target allows, but never beyond the
  trained length nor beyond what the KV cache can fit in memory;
- the reply length is bounded by the context.
"""

from __future__ import annotations

import os
import struct
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional

from .engine import build_plan, detect_hardware
from .types import ExecutionPlan, HardwareProfile, SamplingOptions

DEFAULT_TEMPERATURE = 0.7
DEFAULT_TOP_K = 40
DEFAULT_TOP_P = 0.9
DEFAULT_MAX_TOKENS = 2048
TARGET_CONTEXT_SIZE = 16384
MIN_CONTEXT_SIZE = 2048
SAFETY_MARGIN_MB = 1024


# ---------------------------------------------------------------------------
# GGUF header (metadata only, no tensors)
# ---------------------------------------------------------------------------

_SCALARS = {0: "<B", 1: "<b", 2: "<H", 3: "<h", 4: "<I", 5: "<i",
            6: "<f", 7: "<?", 10: "<Q", 11: "<q", 12: "<d"}


def _read_str(f) -> str:
    (n,) = struct.unpack("<Q", f.read(8))
    return f.read(n).decode("utf-8", errors="replace")


def _read_value(f, vtype: int) -> Any:
    if vtype == 8:
        return _read_str(f)
    if vtype == 9:
        (etype,) = struct.unpack("<I", f.read(4))
        (count,) = struct.unpack("<Q", f.read(8))
        if etype in (8, 9):
            for _ in range(count):
                _read_value(f, etype)
        else:
            f.seek(struct.calcsize(_SCALARS[etype]) * count, 1)
        return None  # arrays are skipped, like the C# reader's ArrayMarker
    fmt = _SCALARS.get(vtype)
    if fmt is None:
        raise ValueError(f"Unknown GGUF value type {vtype}.")
    (v,) = struct.unpack(fmt, f.read(struct.calcsize(fmt)))
    return v


def read_gguf_metadata(model_path: str) -> Dict[str, Any]:
    """Key/value metadata of a GGUF file (array values are omitted)."""
    kv: Dict[str, Any] = {}
    with open(model_path, "rb") as f:
        if f.read(4) != b"GGUF":
            raise ValueError("Not a GGUF file.")
        _version, _tensors, count = struct.unpack("<IQQ", f.read(20))
        for _ in range(count):
            key = _read_str(f)
            (vtype,) = struct.unpack("<I", f.read(4))
            value = _read_value(f, vtype)
            if value is not None:
                kv[key] = value
    return kv


# ---------------------------------------------------------------------------
# Model traits
# ---------------------------------------------------------------------------

@dataclass(frozen=True)
class ModelTraits:
    architecture: str
    trained_context_length: int
    layer_count: int
    head_count_kv: int
    key_length: int
    value_length: int
    sliding_window: int
    file_size_bytes: int
    recommended_temperature: Optional[float] = None
    recommended_top_k: Optional[int] = None
    recommended_top_p: Optional[float] = None

    @property
    def kv_bytes_per_token(self) -> int:
        """Upper-bound KV-cache bytes per context token (16-bit, all layers)."""
        return 2 * self.layer_count * self.head_count_kv * (self.key_length + self.value_length)

    @staticmethod
    def read(model_path: str) -> "ModelTraits":
        kv = read_gguf_metadata(model_path)
        arch = kv.get("general.architecture", "unknown")

        def i(key: str) -> Optional[int]:
            v = kv.get(key)
            return int(v) if isinstance(v, (int, bool)) and not isinstance(v, bool) else None

        def fl(key: str) -> Optional[float]:
            v = kv.get(key)
            return float(v) if isinstance(v, (int, float)) and not isinstance(v, bool) else None

        head_count = i(f"{arch}.attention.head_count") or 0
        head_count_kv = i(f"{arch}.attention.head_count_kv") or head_count
        embedding = i(f"{arch}.embedding_length") or 0
        head_dim = embedding // head_count if head_count > 0 else 0
        key_length = i(f"{arch}.attention.key_length") or head_dim
        value_length = i(f"{arch}.attention.value_length") or key_length
        top_k = i("general.sampling.top_k")
        return ModelTraits(
            architecture=arch,
            trained_context_length=i(f"{arch}.context_length") or 0,
            layer_count=i(f"{arch}.block_count") or 0,
            head_count_kv=head_count_kv,
            key_length=key_length,
            value_length=value_length,
            sliding_window=i(f"{arch}.attention.sliding_window") or 0,
            file_size_bytes=os.path.getsize(model_path),
            recommended_temperature=fl("general.sampling.temp"),
            recommended_top_k=top_k if top_k is not None and top_k > 0 else None,
            recommended_top_p=fl("general.sampling.top_p"),
        )


# ---------------------------------------------------------------------------
# Configuration
# ---------------------------------------------------------------------------

@dataclass(frozen=True)
class AutoConfiguration:
    hardware: HardwareProfile
    plan: ExecutionPlan
    model: ModelTraits
    sampling: SamplingOptions
    context_size: int
    max_tokens: int
    notes: List[str] = field(default_factory=list)


def _fmt(v: Optional[float]) -> str:
    return "-" if v is None else f"{v:.2f}".rstrip("0").rstrip(".")


def compute_configuration(hw: HardwareProfile, plan: ExecutionPlan, model: ModelTraits) -> AutoConfiguration:
    """The pure decision, separated from probing so it can be tested."""
    notes = [
        f"backend {plan.backend.name}, {plan.thread_count} threads, RAM budget {plan.ram_budget_mb} MB "
        f"(cpu threads {hw.cpu_threads}, cuda devices {hw.cuda_device_count}, free RAM {hw.ram_free_mb} MB)"
    ]

    context = TARGET_CONTEXT_SIZE
    if 0 < model.trained_context_length < context:
        context = model.trained_context_length
        notes.append(f"context limited to the trained length {model.trained_context_length}")

    budget_mb = plan.ram_budget_mb if plan.ram_budget_mb > 0 else hw.ram_free_mb
    model_mb = max(0, model.file_size_bytes) // (1024 * 1024)
    kv_per_token = model.kv_bytes_per_token
    if budget_mb > 0 and kv_per_token > 0:
        free_for_kv_mb = budget_mb - model_mb - SAFETY_MARGIN_MB if budget_mb > model_mb + SAFETY_MARGIN_MB else 0
        by_memory = free_for_kv_mb * 1024 * 1024 // 2 // kv_per_token
        if by_memory < context:
            context = max(MIN_CONTEXT_SIZE, by_memory)
            notes.append(f"context limited by memory: ~{kv_per_token // 1024} KB of KV cache per token, "
                         f"{free_for_kv_mb} MB available after the model")
    if context > MIN_CONTEXT_SIZE:
        context -= context % 1024
    context = max(MIN_CONTEXT_SIZE, int(context))

    max_tokens = min(DEFAULT_MAX_TOKENS, context // 4)

    sampling = SamplingOptions(temperature=DEFAULT_TEMPERATURE, top_k=DEFAULT_TOP_K, top_p=DEFAULT_TOP_P)
    if (model.recommended_temperature is not None or model.recommended_top_p is not None
            or model.recommended_top_k is not None):
        notes.append(
            f"the model file suggests temp={_fmt(model.recommended_temperature)} "
            f"top-k={model.recommended_top_k if model.recommended_top_k is not None else '-'} "
            f"top-p={_fmt(model.recommended_top_p)}; defaults used: "
            f"temp={DEFAULT_TEMPERATURE} top-k={DEFAULT_TOP_K} top-p={DEFAULT_TOP_P}")

    notes.append(f"context {context} tokens, replies up to {max_tokens} tokens")
    return AutoConfiguration(hw, plan, model, sampling, context, max_tokens, notes)


def auto_configure(model_path: str, overrides: Optional[ExecutionPlan] = None) -> AutoConfiguration:
    """Configure ``model_path`` for the current machine."""
    traits = ModelTraits.read(model_path)
    hw = detect_hardware()
    plan = build_plan(model_path, overrides)
    return compute_configuration(hw, plan, traits)

File: src/DesireeIA/AutoConfig.cs
// DesireeIA
// Copyright (c) Passaro Francesco Paolo. All rights reserved.
// Licensed under the DesireeIA License - see LICENSE and the "License"
// section of README.md for full terms: no modification, no unauthorized
// integration, no AI training/ingestion without explicit written consent
// from the author.

using DesireeIA.Formats;

namespace DesireeIA;

/// <summary>
/// What the auto-configuration needs to know about a model, read from the
/// GGUF header only (no weights are loaded).
/// </summary>
public sealed record ModelTraits(
    string Architecture,
    uint TrainedContextLength,
    uint LayerCount,
    uint HeadCountKv,
    uint KeyLength,
    uint ValueLength,
    uint SlidingWindow,
    long FileSizeBytes,
    float? RecommendedTemperature,
    int? RecommendedTopK,
    float? RecommendedTopP)
{
    /// <summary>
    /// Estimated KV-cache bytes per context token (keys + values, 16-bit, every
    /// layer). An upper bound: sliding-window layers and latent/compressed
    /// attention (DeepSeek MLA) need less, so budgets computed from it err on
    /// the safe side.
    /// </summary>
    public ulong KvBytesPerToken => 2UL * LayerCount * HeadCountKv * (KeyLength + ValueLength);

    /// <summary>Reads the traits of a GGUF model file.</summary>
    public static ModelTraits Read(string modelPath)
    {
        var reader = new GgufReader();
        var meta = reader.Open(modelPath);
        var kv = reader.RawKeyValues;
        var arch = meta.Architecture;

        uint headCount = U(kv, $"{arch}.attention.head_count") ?? 0;
        uint headCountKv = U(kv, $"{arch}.attention.head_count_kv") ?? headCount;
        // Without explicit key/value lengths the head size is embedding / heads.
        uint headDim = headCount > 0 ? meta.EmbeddingSize / headCount : 0;
        uint keyLength = U(kv, $"{arch}.attention.key_length") ?? headDim;
        uint valueLength = U(kv, $"{arch}.attention.value_length") ?? keyLength;

        // general.sampling.* are the author's suggested values (GGUF spec).
        // top_k <= 0 means "disabled" there, so it's reported as no suggestion.
        int? topK = I(kv, "general.sampling.top_k");
        return new ModelTraits(
            Architecture: arch,
            TrainedContextLength: meta.ContextLength,
            LayerCount: meta.LayerCount,
            HeadCountKv: headCountKv,
            KeyLength: keyLength,
            ValueLength: valueLength,
            SlidingWindow: U(kv, $"{arch}.attention.sliding_window") ?? 0,
            FileSizeBytes: new FileInfo(modelPath).Length,
            RecommendedTemperature: F(kv, "general.sampling.temp"),
            RecommendedTopK: topK is > 0 ? topK : null,
            RecommendedTopP: F(kv, "general.sampling.top_p"));
    }

    private static uint? U(IReadOnlyDictionary<string, object> kv, string key) =>
        kv.TryGetValue(key, out var v) ? v switch
        {
            uint u => u,
            int i when i >= 0 => (uint)i,
            ulong ul when ul <= uint.MaxValue => (uint)ul,
            long l when l >= 0 && l <= uint.MaxValue => (uint)l,
            ushort us => us,
            byte b => b,
            _ => null
        } : null;

    private static int? I(IReadOnlyDictionary<string, object> kv, string key) =>
        kv.TryGetValue(key, out var v) ? v switch
        {
            int i => i,
            uint u when u <= int.MaxValue => (int)u,
            long l when l >= int.MinValue && l <= int.MaxValue => (int)l,
            ulong ul when ul <= int.MaxValue => (int)ul,
            _ => null
        } : null;

    private static float? F(IReadOnlyDictionary<string, object> kv, string key) =>
        kv.TryGetValue(key, out var v) ? v switch
        {
            float f => f,
            double d => (float)d,
            _ => null
        } : null;
}

/// <summary>
/// A complete, ready-to-use configuration for a model on THIS machine:
/// execution plan (backend, threads, RAM), sampling, context and output
/// length. <see cref="Notes"/> explains every choice in plain words.
/// </summary>
public sealed record AutoConfiguration(
    HardwareProfile Hardware,
    ExecutionPlan Plan,
    ModelTraits Model,
    SamplingOptions Sampling,
    int ContextSize,
    int MaxTokens,
    IReadOnlyList<string> Notes);

/// <summary>
/// Picks the parameters a model should run with on the current hardware, so
/// that loading a model "just works" without hand-tuning:
/// <list type="bullet">
/// <item>backend, threads and RAM budget come from the native planner
/// (<see cref="DesireeIAEngine.BuildPlan"/>), which probes CPU features,
/// CUDA/Intel/Metal devices and free memory;</item>
/// <item>sampling is never greedy: greedy decoding falls into repetitive loops
/// on long generations, the main cause of "the local model repeats itself";</item>
/// <item>the context is as large as the target allows, but never beyond what
/// the model was trained on nor beyond what the KV cache can fit in memory;</item>
/// <item>the output length is bounded by the context, so a reply can never
/// crowd out the prompt.</item>
/// </list>
/// </summary>
public static class AutoConfigurator
{
    /// <summary>Default temperature: varied but focused text.</summary>
    public const float DefaultTemperature = 0.7f;
    public const int DefaultTopK = 40;
    public const float DefaultTopP = 0.9f;
    /// <summary>Longest reply by default.</summary>
    public const int DefaultMaxTokens = 2048;
    /// <summary>Context to aim for: room for a system prompt with tool
    /// descriptions, a conversation and a full-length reply.</summary>
    public const int TargetContextSize = 16384;
    /// <summary>Smallest context worth running with.</summary>
    public const int MinContextSize = 2048;
    /// <summary>Memory always left to the OS and the rest of the process.</summary>
    public const ulong SafetyMarginMb = 1024;

    /// <summary>
    /// Configures <paramref name="modelPath"/> for the current machine.
    /// <paramref name="overrides"/> works as in <see cref="DesireeIAEngine.BuildPlan"/>.
    /// </summary>
    public static AutoConfiguration Configure(string modelPath, ExecutionPlan? overrides = null)
    {
        var traits = ModelTraits.Read(modelPath);
        var hw = DesireeIAEngine.DetectHardware();
        var plan = DesireeIAEngine.BuildPlan(modelPath, overrides);
        return Compute(hw, plan, traits);
    }

    /// <summary>The pure decision, separated from probing so it can be tested.</summary>
    public static AutoConfiguration Compute(HardwareProfile hw, ExecutionPlan plan, ModelTraits model)
    {
        var notes = new List<string>
        {
            $"backend {plan.Backend}, {plan.ThreadCount} threads, RAM budget {plan.RamBudgetMb} MB " +
            $"(cpu threads {hw.CpuThreads}, cuda devices {hw.CudaDeviceCount}, free RAM {hw.RamFreeMb} MB)"
        };

        // ── Context ──────────────────────────────────────────────────────
        int context = TargetContextSize;
        if (model.TrainedContextLength > 0 && model.TrainedContextLength < context)
        {
            context = (int)model.TrainedContextLength;
            notes.Add($"context limited to the trained length {model.TrainedContextLength}");
        }

        ulong budgetMb = plan.RamBudgetMb > 0 ? plan.RamBudgetMb : hw.RamFreeMb;
        ulong modelMb = (ulong)Math.Max(0, model.FileSizeBytes) / (1024 * 1024);
        ulong kvPerToken = model.KvBytesPerToken;
        if (budgetMb > 0 && kvPerToken > 0)
        {
            // Weights may sit in VRAM on GPU backends, but VRAM size isn't
            // probed: counting them against RAM keeps the estimate safe.
            ulong freeForKvMb = budgetMb > modelMb + SafetyMarginMb ? budgetMb - modelMb - SafetyMarginMb : 0;
            // Half of what is left: the KV cache is not the only buffer.
            ulong byMemory = freeForKvMb * 1024 * 1024 / 2 / kvPerToken;
            if (byMemory < (ulong)context)
            {
                context = (int)Math.Max((ulong)MinContextSize, byMemory);
                notes.Add($"context limited by memory: ~{kvPerToken / 1024} KB of KV cache per token, " +
                          $"{freeForKvMb} MB available after the model");
            }
        }
        if (context > MinContextSize) context -= context % 1024;
        context = Math.Max(MinContextSize, context);

        // ── Output length ────────────────────────────────────────────────
        int maxTokens = Math.Min(DefaultMaxTokens, context / 4);

        // ── Sampling ─────────────────────────────────────────────────────
        var sampling = new SamplingOptions
        {
            Temperature = DefaultTemperature,
            TopK = DefaultTopK,
            TopP = DefaultTopP,
        };
        if (model.RecommendedTemperature is not null || model.RecommendedTopP is not null || model.RecommendedTopK is not null)
        {
            notes.Add("the model file suggests " +
                      $"temp={Fmt(model.RecommendedTemperature)} top-k={(model.RecommendedTopK?.ToString() ?? "-")} " +
                      $"top-p={Fmt(model.RecommendedTopP)}; defaults used: " +
                      $"temp={DefaultTemperature} top-k={DefaultTopK} top-p={DefaultTopP}");
        }

        notes.Add($"context {context} tokens, replies up to {maxTokens} tokens");
        return new AutoConfiguration(hw, plan, model, sampling, context, maxTokens, notes);
    }

    private static string Fmt(float? f) =>
        f?.ToString("0.##", System.Globalization.CultureInfo.InvariantCulture) ?? "-";
}
